hal/include/hal/can_api.h

```diff
@@ -63,7 +63,7 @@ typedef struct {
     int td_function;
 } can_pinmap_t;
 
-typedef void (*can_irq_handler)(uint32_t id, CanIrqType type);
+typedef void (*can_irq_handler)(uintptr_t id, CanIrqType type);
```

Comment:
This carries the additional meaning that the id is a pointer. `size_t` may be more appropriate (and passing in the index in the table rather than a pointer).

Reply:
My understanding of how the drivers CAN Class and the HAL can_api interoperate is that the CAN constructor passes the address of itself cast to an integer type via `can_irq_init`, which becomes the id in the table. CAN has a static method `_irq_handler` which takes the id/address and then casts it back to a `CAN *` in order to call the correct callbacks. 

As this `_irq_handler` method is what the `can_irq_handler` type is used to refer to in all the `can_api` implementations, I was intentionally using a pointer type because the id is indeed a pointer - it's pointers all the way down if you will. But maybe I am peeling back the curtain too much here and `size_t` would be more appropriate (?)

Reply:
I had a deeper look at a few hal impl and as far as I can see the ID is never used in the lowlevel driver, it is only passed back to the callbacks.

The fact that this ID is a pointer instead of an plain integer is a shortcut that the Driver takes in the definition/management of IDs. By changing the type of the HAL here we are bending the definition of ID from the HAL to accommodate the hacky thing the Driver does (casting its `this` to something not guaranteed to be able to hold its value).

MCU-Driver-HAL has similar needs and made the choice to make it an `intptr_t` with the nuance that it's not called an ID but a context. (id also to some extend imply unicity which is not even required by the driver itself).

So if we intend to keep the pointer shortcut (which makes sense for efficiency), **I recommand we use "context" rather than "id"**. This way we have a clear "pointer to the context related to that callback" rather than "and Id that happen to also be a pointer so we changed its type to make sure it cal hold that pointer's value".

Reply:
This makes sense - I'll amend with this naming change, for greater clarity. Is there a preference between `intptr_t` and `uintptr_t` here?

Reply:
No strong preference for me. If sign does not add value (e.g. as in the distance between two pointers) I'd probably use the unsigned version but if someone thinks otherwise I'm ok with either.

Reply:
> So if we intend to keep the pointer shortcut (which makes sense for efficiency), I recommand we use "context" rather than "id". This way we have a clear "pointer to the context related to that callback" rather than "and Id that happen to also be a pointer so we changed its type to make sure it cal hold that pointer's value".

Revisiting, this makes even more sense - code currently uses `id` to refer to both the context itself and the index into the array used to select the context. Now we have a clear distinction.

Reply:
Rebased with naming changes :)